### PR TITLE
Image Acquisition Leiden

### LINKED ISSUE #26: Implementation of TSE
- [x] Constructor for TSE sequence
- [x] Test unrolling, measure performance
- [x] Construct experiment
- [ ] Implement reconstruction for TSE
- [ ] Execute TSE sequence at Berlin scanner

### LINKED ISSUE #28: Append to existing log file in session

### LINKED ISSUE #30: BUG: Enabling more then two channels is not possible
- When more than two channels are enabled in the device configuration, the read number of enabled channels from card equals zero 
- Happens directly after channel enable: `sp.spcm_dwSetParam_i32(self.card, sp.SPC_CHENABLE, channel_selection)`
- Enabling channel 0 and/or channel 1 works, even independently

### Resolution attempts
- Tried to replaced `channel_selection` by `sp.CHANNEL0 | sp.CHANNEL1 | sp.CHANNEL2` and observed the same behavior
- Tried to set the channel amplitude to `10000` according to the other channels

### LINKED ISSUE #32: Add system settings to meta information of acquisition data
- [ ] Parameter of `TxCard`: Output level, filter type, sample rate
- [ ] Parameter of `RxCard`:
- [ ] Parameter of `SequenceProvider`: rf to volt scaling, gpa scaling, gradient coefficients, enabled channels, impedance, max. amplitudes

### LINKED ISSUE #34: Example from readme does not run
1. An import has to be added `from console.spcm_control.interface_acquisition_data import AcquisitionData`
2. `fov_offset` is not yet defined in `AcquisitionParameter`

### LINKED ISSUE #36: Use Ruff for linting
https://docs.astral.sh/ruff/

### PR DESCRIPTION
**Pull request by schote**
- [x] Implementation of TSE Sequence 2D/3D
- [x] Update gradient scaling by hardware parameter: coil sensitivity, gpa gain
- [x] Add image reconstruction algorithm
- [x] Update sequence plot
- [x] Improvement of rollout performance
- [x] Switch to ruff for linting
- [x] Enable execution of experiment scripts from console

closes #34 
closes #32 
closes #28 
closes #26 
closes #36 
closes #30

### CODE REVIEW

File: experiments/freq_calibration.py
Comment (berksilemek) [inline experiments/freq_calibration.py:1]: Should we also track the larmor frequency and room temperature by date?

File: experiments/reco/tx_adjust_fit.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+from scipy.optimize import curve_fit
+
+# %%
+session_name = "2023-11-01-session"
+acquisition = "2023-11-01-093257-tx_adjust_fid"
+
+# session_name = "2023-10-31-session"
+# acquisition = "2023-10-31-172257-tx_adjust_fid"
+
+raw_data = np.load(f"/home/schote01/spcm-console/{session_name}/{acquisition}/raw_data.npy")
```
Comment (berksilemek) [inline experiments/reco/tx_adjust_fit.py:14]: It could be fine if we could eliminate user-specific path definitions

File: experiments/tse.py [hunk 2]
```diff
@@ -1,3 +1,120 @@
...
+    fov=Dimensions(x=220e-3, y=220e-3, z=225e-3),
+    # n_enc=Dimensions(x=64, y=64, z=0)
+    n_enc=Dimensions(x=128, y=128, z=0)
+)
+
+# %%
+# Plot sequence section
+acq.seq_provider.from_pypulseq(seq)
+seq_unrolled = acq.seq_provider.unroll_sequence(larmor_freq=2e6, grad_offset=Dimensions(0, 0, 0))
+fig, ax = plot_unrolled_sequence(seq_unrolled, seq_range=(0, 640000), output_limits=[200, 6000, 6000, 6000])
```
Comment (berksilemek) [inline experiments/tse.py:37]: Is it possible to read function parameters from the current settings? Maybe todo?

File: src/console/spcm_control/rx_device.py [hunk 3]
```diff
@@ -320,14 +371,14 @@ def _gated_timestamps_stream(self):
                         else:
                             gate_data = rx_data[index_0 : index_0 + int(total_bytes / 2)]
 
-                        pre_trigger_cut = (self.pre_trigger - 1) * self.num_channels.value
```
Comment (berksilemek) [inline src/console/spcm_control/rx_device.py:323]: I forgot why we deleted -1. I should doublecheck this part.

**Review by berksilemek — APPROVED** [review]

File: experiments/freq_calibration.py
Comment (schote) [reply experiments/freq_calibration.py:1]: At the moment you can add any information manually. To track true frequency and temperature automatically I would add automatic frequency calibration (which is work in progress) then we can add it as a standard routine. The frequency used for excitation is already stored within the acquisition parameters. For temperature tracking I would add some sensors.

File: experiments/reco/tx_adjust_fit.py (hunk 1 above)
Comment (schote) [reply experiments/reco/tx_adjust_fit.py:14]: Yes, I agree, the problem with these reco scripts is, that we don't have a general data store which is accessible by anyone (I would prefer to keep the data per user). But maybe the best solution would be to keep all reco stuff out of this repository?

File: experiments/tse.py (hunk 2 above)
Comment (schote) [reply experiments/tse.py:37]: I'm not sure I understand what you mean, this is just to plot the unrolled sequence. We could seperate unrolling and execution of a sequence. Then one could set a sequence in the acquisition control and plot exactly the same sequence that is replayed?

File: src/console/spcm_control/rx_device.py (hunk 3 above)
Comment (schote) [reply src/console/spcm_control/rx_device.py:323]: I think this was just a test from my side, without the -1 it should be fine. 
But we should definetly double-check this in general. I suggest to solve this in PR #29

File: experiments/tse.py (hunk 2 above)
Comment (berksilemek) [reply experiments/tse.py:37]: In this part: ...output_limits=[200, 6000, 6000, 6000], it seemed to me we unnecessarly set the output limits again. Is not it already set in the project settings globally? Do we need to specify them again? I could not follow this part.

Comment (schote) [reply experiments/tse.py:37]: This is only to plot the unrolled sequence with the correct output limits. The unrolled sequence is in int16 values.
We could add the plot function to the sequence provider, then we don't need to pass them again.

**Pull request merged**